src/Build/Graph/ProjectGraph.cs

```diff
@@ -175,12 +175,25 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
             {
                 PropertyDictionary<ProjectPropertyInstance> globalPropertyDictionary = CreatePropertyDictionary(entryPoint.GlobalProperties);
                 var configurationMetadata = new ConfigurationMetadata(FileUtilities.NormalizePath(entryPoint.ProjectFile), globalPropertyDictionary);
-                ProcessNode(configurationMetadata, nodeStates, projectCollection, globalPropertyDictionary);
+                if(!nodeStates.TryGetValue(configurationMetadata, out NodeState nodeState))
+                {
+                    ProcessNode(configurationMetadata, nodeStates, projectCollection, globalPropertyDictionary);
+                }
                 entryPointNodes.Add(_allParsedProjects[configurationMetadata]);
             }
 
+            var graphRoots = new List<ProjectGraphNode>();
```

Comment:
preallocate as `new List<ProjectGraphNode>(entryPointNodes.Count)` to possibly save a couple of resizes? Generally good practice, though I doubt this path will be hot enough to ever really matter.